### ISSUE TITLE
Behavior of broken Sage package is confusing.

### ISSUE THREAD

**Issue by Zimmi48** [NixOS/nixpkgs#55648]
## Issue description

cc @timokau

I tried to install `sage` using the attribute in nixpkgs-unstable, it fetched a bunch of stuff, then ran tests for hours until it failed in the following way:

```
----------------------------------------------------------------------
sage -t --long /nix/store/1q8yyxdl943mx6xjrpfrsmpxayqb6162-sage-src-8.4/src/sage/modular/modform/element.py  # Timed out
sage -t --long /nix/store/1q8yyxdl943mx6xjrpfrsmpxayqb6162-sage-src-8.4/src/sage/symbolic/constants.py  # Timed out after testing finished
----------------------------------------------------------------------
Total time for all tests: 25598.8 seconds
    cpu time: 9816.4 seconds
    cumulative wall time: 12004.7 seconds
builder for '/nix/store/vv5ka0i2h5rjwycqlck9rkzwdk6xz6dh-sage-tests-8.4.drv' failed with exit code 4
cannot build derivation '/nix/store/2kh547n03xdshrny8g7fhdyvfcsh5gm2-sage-8.4.drv': 1 dependencies couldn't be built
error: build of '/nix/store/2kh547n03xdshrny8g7fhdyvfcsh5gm2-sage-8.4.drv' failed
```

After looking into the nixpkgs sources, I understood that this test package should have been built on Hydra but it was not available because it was broken. Said otherwise, I should have deduced that since the tests were running, I should kill the build and upgrade my nixpkgs version. This is quite confusing and it would be great if there was a way to make this more intuitive. But I have no idea how to fix this, except by quickly adding `broken=true` to the derivation when it fails on Hydra, without waiting for a fix.

### Steps to reproduce

`nix-env -f '<nixpkgs>' -iA sage` with nixpkgs pointing to 7c75afaf1ab983091c9064576eaf793594f01058.

**Comment by timokau** [comment]
I agree that this is a problem. It would be nice if substituters like hydra could report cache failures. Even better if packages are marked as broken automatically.

It is not a sage specific problem though. You can try starting a discussion about it on [discourse](https://discourse.nixos.org/). For what it's worth, the sage testsuite passes on current master and the build is cached on nixos-unstable-small. Unfortunately the regular unstable channel hasn't advanced yet.

**Comment by Zimmi48** [comment]
Done: https://discourse.nixos.org/t/behavior-of-undeclared-broken-packages-is-confusing/2123

**Comment by matthewbauer** [comment]
This seems like it is at least partly on sage. We shouldn't be forcing everyone to run tests that last this long. I would say that about 15 minutes should be the upper bound for a test.

It's also a little weird to put the tests in a separate derivation. Usually these are put in the checkPhase of the derivation it's testing. I know there are some proposals to run these in a separate derivation, but even then I think we want it contained within the checkPhase of the derivation. Otherwise it's not clear what the test is actually testing (and has the weird "fake" phases in the ceck derivation).

**Comment by timokau** [comment]
> We shouldn't be forcing everyone to run tests that last this long

We're not forcing anyone to do that any more than we're forcing anyone to build chromium. The tests are normally run by hydra.

> I would say that about 15 minutes should be the upper bound for a test.

That just isn't enough for sage.

> It's also a little weird to put the tests in a separate derivation.

Its a little similar to @Profpatsch's "with tests". Its done because the build and the test both take significant time and its not always necessary to rebuild sage when rerunning the tests (for example when adjusting the runtime env). Also makes it possible to disable the tests while still using substitution or to just run a single test (mostly for `git-bisect`).

Its a bit ad-hoc here, but I think this is the way to go for install tests.

**Comment by matthewbauer** [comment]
Chromium has to be built by someone but no one has to run these tests. I was hoping there would be a way to disable the long running tests of sage to keep the time reasonable. If not, let’s not make hydra run them either, because we don’t have unlimited computing power. How about putting them in passthru.tests and just have developers run them? This would avoid having to rebuild on every change.

**Comment by timokau** [comment]
But the main point of the testsuite is to detect regressions caused by changes to other packages. Sage is heavily relying on a lot of packages in the math ecosystem. Just having me run the test before updates is of little use.

I think ~1h (https://hydra.nixos.org/build/88795195#tabs-buildsteps) of time is entirely reasonable.

**Comment by spacekitteh** [comment]
Still important.
